platform/logger_internal.h

```diff
@@ -0,0 +1,284 @@
+
+#ifndef MBED_CONF_MAX_LOG_STR_SIZE
+#define MBED_CONF_MAX_LOG_STR_SIZE      128
+#endif
+
+#ifndef MBED_CONF_MAX_BUFFER_SIZE
+#define MBED_CONF_MAX_BUFFER_SIZE       512
+#endif
+
+#if defined (MBED_ID_BASED_TRACING)
+#define LOG_DATA_TYPE_   uint32_t
+#else
+#define LOG_DATA_TYPE_   char
+#endif
+
+#if defined(__ARMCC_VERSION)
+#define LOG_FILE_NAME_                  __MODULE__
+#else
+#define LOG_FILE_NAME_                  __BASE_FILE__
+#endif
+
+#define LOG_SINGLE_STR_SIZE_            MBED_CONF_MAX_LOG_STR_SIZE
+#define LOG_SINGLE_HELPER_STR_SIZE_     LOG_SINGLE_STR_SIZE_
+#define LOG_BUF_SIZE_                   MBED_CONF_MAX_BUFFER_SIZE
+
+// Log-Level Strings
+#define LOG_GEN_                 "GEN "
+#define LOG_ERR_CRITICAL_        "CRT "
+#define LOG_ERR_                 "ERR "
+#define LOG_WARN_                "WARN"
+#define LOG_DEBUG_               "DBG "
+#define LOG_INFO_                "INFO"
+#define LOG_TRACE_               "TRAC"
+
+#define GET_LOG_STRING(ll)      ((ll > LOG_LEVEL_DEBUG) ? LOG_TRACE_ : \
+                                ((ll > LOG_LEVEL_INFO)  ? LOG_DEBUG_ : \
+                                ((ll > LOG_LEVEL_WARN)  ? LOG_INFO_  : \
+                                ((ll > LOG_LEVEL_ERR)   ? LOG_WARN_  : \
+                                ((ll > 1)   ? LOG_ERR_   : LOG_GEN_)))))
+
+#define GET_MODULE_ID_(a, b, c, d)   (375 + (a << 1) + (b << 2) + (c >> 2) + (d >> 1))
+#define GET_MODULE_ID(x)        (GET_MODULE_ID_(x[0], x[1], x[2], x[3]))
+
+#define SET_MODULE(x)           (GET_MODULE_ID(x) & 0xFF)
+#define SET_COUNTER(y)          ((y & 0xFF) << 8)
+#define SET_LINE_NUM(z)         ((z & 0xFFFF) << 16)
+#define TRACE_ID_(x,y,z)        (SET_MODULE(x) | SET_COUNTER(y) | SET_LINE_NUM(z))
+
+// Macros to log ID based data
+#define MBED_LOG_ID_4(...)                            log_id_data(__VA_ARGS__)
+// Data dumped in special section is : Unique ID, format string length and format string
+#if 0
+#define MBED_LOG_ID_3(counter, id, args, fmt, ...)    ({volatile static const __attribute__((section(".keep.log_data"))) char str##counter[] = fmt; \
+                                                       volatile static const __attribute__((section(".keep.log_data"))) uint32_t len##counter = MBED_STRLEN(fmt); \
+                                                       volatile static const __attribute__((section(".keep.log_data"))) uint32_t c##counter = id; \
+                                                       MBED_LOG_ID_4(args, id, ##__VA_ARGS__); \
+                                                      })
+#else
+#define MBED_LOG_ID_3(counter, id, args, fmt, ...)    MBED_LOG_ID_4(args, id, ##__VA_ARGS__);
+#endif
+
+#define MBED_LOG_ID_2(counter, id, ...)               MBED_LOG_ID_3(counter, id, MBED_COUNT_VA_ARGS(__VA_ARGS__), ##__VA_ARGS__)
+#define MBED_LOG_ID_1(mod, fmt, ll, f, l, c, ...)     MBED_LOG_ID_2(c, TRACE_ID_(mod,c,l), "[" mod "][" f "][" MBED_STRINGIFY(l) "]: " fmt, ##__VA_ARGS__)
```

Comment:
Would it be possible to have a compile time flags that drop references to `fmt` ? 
For some users, the file and the line are sufficient to get the origin of the log.